en/replacing-docker-compose.md

```diff
@@ -0,0 +1,103 @@
+- Along with services being project specific, their data should also be project specific.
+
+These were the exact problems [services-flake](https://community.flake.parts/services-flake) was designed to solve. Along with running services natively, it also [integrates with your project's `flake.nix`](https://community.flake.parts/services-flake/start).
+
+>[!info] How are processes managed in services-flake?
+> In NixOS they are managed by [systemd](https://en.wikipedia.org/wiki/Systemd). In services-flake, we use [process-compose](https://github.com/F1bonacc1/process-compose) whose configuration is managed by [process-compose-flake](https://community.flake.parts/process-compose-flake).
+
+{#let-s-get-started}
+
+## Let's get started
+
+Now that we have all the answers. It's time to replace [docker-compose in NY](https://github.com/nammayatri/nammayatri/blob/f056bb994fbf9adefa454319032ca35c34ea65bc/Backend/nix/arion-configuration.nix) with services-flake. We will only look at a few services to keep it simple, for more details, see [PR](https://github.com/nammayatri/nammayatri/pull/3718).
+
+:::{.center}
+![[ny-services-flake.png]]
+:::
+
+{#postgresql}
+
+### PostgreSQL
+
+NY uses about 3 instances of postgresql databases.
+
+One of them is [exported by passetto](https://github.com/nammayatri/passetto/blob/nixify/process-compose.nix) (passetto is a haskell application that encrypts data before storing in postgres), and using it looks like:
+
+```nix
+{
+  services.passetto.enable = true;
+}
+```
+
+By using [[modules|module system]], we can hide the implementation details and only expose the `passetto` service to the user, and use it as above.
+
+The other two instances are used by the [postgres-with-replica module](https://github.com/nammayatri/nammayatri/blob/ccab8da607cfd8d4e9f7d28b55b83e22eec1af9b/Backend/nix/services/postgres-with-replica.nix). This module starts two services (`primary` and `replica` database) and a [pg-basebackup](https://www.postgresql.org/docs/current/app-pgbasebackup.html) process (to sync `replica` with `primary` during init). For the user it is:
+
+```nix
+{
+  services.postgres-with-replica.enable = true;
+}
+```
+
+{#redis}
+
+### Redis
+
+Redis and its clustered version are pretty straightforward:
+
+```nix
+{
+  services.redis.enable = true;
+  services.redis-cluster.enable = true;
+}
+```
+
+{#cool-things}
+
+## Cool things
+
+By not depending on docker anymore, we can now run the entire NY backend with one command and its all defined in a [single place](https://github.com/nammayatri/nammayatri/blob/ccab8da607cfd8d4e9f7d28b55b83e22eec1af9b/Backend/nix/services/nammayatri.nix). 
+
+That is not all, we can also reuse this to do much more, like defining [loadtest](https://github.com/nammayatri/nammayatri/blob/ccab8da607cfd8d4e9f7d28b55b83e22eec1af9b/Backend/load-test/default.nix) config and run it in CI/local, again, with one command.
```

Comment:
Is it in scope for this blog post to expand on the loadtest? In particular, emphasizing on how users can *share* services-flake modules, and then build on top of them. Here, we take the module to run entire NY stack, and then extend it to add a bunch of load-test processes, before bringing the whole thing end (as the load-test ends).

Reply:
I think it's a good idea to mention load-test because it is one of the things we are able to do as a result of adding services-flake. If that complicates things, I can probably remove it and talk about it in another post explaining how you can reuse services, courtesy of flake-parts.

Reply:
You could add a sentence or two introducing the users about it, and then dedicate another post on it as you see fit.

Reply:
https://github.com/nixos-asia/website/pull/28/commits/b0a408f06ca82e26f34ba404ab48eebb6a662140

Reply:
Copied the second half from [your comment](https://github.com/nixos-asia/website/pull/28#discussion_r1510357427) to expand on load-test